src/main/java/com/blazemeter/taurus/junit/reporting/TaurusReporter.java

```diff
@@ -0,0 +1,152 @@
+package com.blazemeter.taurus.junit.reporting;
+
+import com.blazemeter.taurus.junit.Utils;
+import org.json.JSONObject;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class TaurusReporter {
+
+    private FileWriter outStream;
+    private static final Logger log = Logger.getLogger(TaurusReporter.class.getName());
+    private final LinkedBlockingQueue<Sample> queue = new LinkedBlockingQueue<>();
+
+    private int activeThreads = 0;
+
+    private volatile boolean isStopped = false;
+    private final SampleFormatter formatter;
+    private final Thread reporter;
+
+    private final boolean isVerbose;
+
+    public TaurusReporter(String fileName) {
+        File file = new File(fileName);
+        try {
+            outStream = new FileWriter(file);
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to open file " + fileName, e);
+        }
+
+        formatter = createFormatter(file);
+        isVerbose = formatter instanceof JSONFormatter;
+        log.info("File: " + fileName + ", formatter: " + formatter.getClass().getSimpleName());
+
+        reporter = new PoolWorker();
+        reporter.setName("Reporter thread");
+        reporter.setDaemon(true);
+        reporter.start();
+    }
+
+    private SampleFormatter createFormatter(File file) {
+        if (Utils.getFileExtension(file).equals("csv")) {
+            CSVFormatter formatter = new CSVFormatter();
+            try {
+                outStream.write(formatter.getHeader());
+                outStream.write("\n");
+            } catch (IOException e) {
+                log.log(Level.WARNING, "Failed to write CSV header", e);
+            }
+            return formatter;
+        }
+        return new JSONFormatter();
+    }
+
+    public void writeSample(Sample sample) {
+        try {
+            queue.put(sample);
+        } catch (InterruptedException e) {
+            log.log(Level.SEVERE, "Failed to put sample in queue: ", e);
+        }
+    }
+
+    public void close() throws Exception {
+        log.info("Closing reporter stream");
+        isStopped = true;
+        reporter.join();
+        outStream.close();
+    }
+
+    private class PoolWorker extends Thread {
+        public void run() {
+            while (!isStopped || !queue.isEmpty()) {
```

Comment:
is `isEmpty()` correct here? Queue will be empty most of times, but the idea of queue is to wait for sample on it...

Reply:
if worker has stopped `isEmpty()` will write the remains of the queue